src/Disks/ObjectStorages/ObjectStorageVFSGCThread.cpp

```diff
@@ -0,0 +1,190 @@
+#include "ObjectStorageVFSGCThread.h"
+#include <ranges>
+#include "Common/ZooKeeper/ZooKeeperLock.h"
+#include "Compression/CompressedReadBuffer.h"
+#include "Compression/CompressedWriteBuffer.h"
+#include "Disks/ObjectStorages/DiskObjectStorageVFS.h"
+#include "Interpreters/Context.h"
+
+namespace DB
+{
+using enum VFSTransactionLogItem::Type;
+
+namespace ErrorCodes
+{
+extern const int LOGICAL_ERROR;
+}
+
+static constexpr std::string_view VFS_SNAPSHOT_PREFIX = "vfs_snapshot_";
+
+ObjectStorageVFSGCThread::ObjectStorageVFSGCThread(DiskObjectStorageVFS & storage_, ContextPtr context)
+    : storage(storage_)
+    , log(&Poco::Logger::get(fmt::format("VFSGC({})", storage_.getName())))
+    , zookeeper_lock(storage.zookeeper, storage.traits.locks_node, "gc_lock")
+{
+    task = context->getSchedulePool().createTask(
+        log->name(),
+        [this]
+        {
+            try
+            {
+                run();
+            }
+            catch (...)
+            {
+                LOG_DEBUG(log, "Task threw an exception, rescheduling");
+                tryLogCurrentException(log, __PRETTY_FUNCTION__);
+            }
+        });
+
+    task->activateAndSchedule();
+}
+
+ObjectStorageVFSGCThread::~ObjectStorageVFSGCThread() = default;
+
+void ObjectStorageVFSGCThread::run()
+{
+    SCOPE_EXIT({
+        zookeeper_lock.unlock();
+        task->scheduleAfter(storage.gc_thread_sleep_ms);
+    });
+
+    if (!zookeeper_lock.tryLock())
+    {
+        LOG_DEBUG(log, "Failed to acquire lock, sleeping");
+        return;
+    }
+
+    const Strings batch = storage.zookeeper->getChildren(storage.traits.log_base_node);
```

Comment:
what if the prev run will fail to remove some childs after creating the snapshot? May be better to store the last_successful_snapshot_position & maybe even something like last_committed_offset

Reply:
Regarding two potential situations where it may break in the GC run:
1. When we write new snapshot successfully but fail to remove the old one from object storage:
  New GC run will have 2 snapshots in log batch. This particular situation can be solved by using the former snapshot and re-applying all changes till latter snapshot name's logpointer.

2. When we remove objects from storage but fail to remove log range. Differs from (1) as we can't retry objects deletion

Reply:
Solved using reconciliation in GC